{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import random\n",
    "from Bio import Entrez, SeqIO\n",
    "from matplotlib import pyplot\n",
    "\n",
    "def getSequencesFromFile(filePath):\n",
    "    \"\"\"\n",
    "    Reads the FASTA file at the specified filePath and returns the sequences within it.\n",
    "\n",
    "    Parameters:\n",
    "        filePath - a path to the file you want to read\n",
    "    Returns:\n",
    "        A list of strings containing each of the sequences in the specied file\n",
    "    \"\"\"\n",
    "    with open(filePath) as handle:\n",
    "        sequenceStrings = []\n",
    "        records = SeqIO.parse(handle, \"fasta\")\n",
    "        for record in records:\n",
    "            sequenceStrings.append(str(record.seq))\n",
    "\n",
    "    return sequenceStrings\n",
    "\n",
    "def getSequencesFromNCBI(databaseName, accessionId):\n",
    "    \"\"\"\n",
    "    Downloads the specified record from the NCBI and returns the sequences within it.\n",
    "\n",
    "    Parameters:\n",
    "        databaseName - the name of the NCBI database to download from, i.e. \"protein\" or \"nucleotide\"\n",
    "        accessionId - the accession number/id of the specified record, i.e. \"NM_000280.4\"\n",
    "    Returns:\n",
    "        A list of strings containing each of the sequences in the specied file\n",
    "    \"\"\"\n",
    "    with Entrez.efetch(db=databaseName, id=accessionId, rettype=\"fasta\", retmode=\"text\") as handle:\n",
    "        sequenceStrings = []\n",
    "        records = SeqIO.parse(handle, \"fasta\")\n",
    "        for record in records:\n",
    "            sequenceStrings.append(str(record.seq))\n",
    "\n",
    "    return sequenceStrings\n",
    "\n",
    "def dotplot(sequenceA, sequenceB, windowSize):\n",
    "    \"\"\"\n",
    "    Creates a dotplot showing substrings in sequenceA and sequenceB that are identical.\n",
    "\n",
    "    Parameters:\n",
    "        sequenceA - a string containing the first sequence to compare\n",
    "        sequenceB - a string containing the second sequence to compare\n",
    "        windowSize - the length of substrings to compare in each sequence\n",
    "    Returns:\n",
    "        Nothing.\n",
    "        However, a dotplot will shown in a new window created by matplotlib,\n",
    "            with position in sequenceA on the x-axis and position in sequenceB on the y-axis.\n",
    "        This function will block until the dotplot window is closed.\n",
    "    \"\"\"\n",
    "    xList = []\n",
    "    yList = []\n",
    "    maxIndexA = len(sequenceA) - windowSize + 1\n",
    "    maxIndexB = len(sequenceB) - windowSize + 1\n",
    "    for indexA in range(maxIndexA):\n",
    "        # Iterate throught every substring of length windowSize in sequenceA\n",
    "        substringA = sequenceA[indexA:indexA+windowSize]\n",
    "        for indexB in range(maxIndexB):\n",
    "            # Iterate throught every substring of length windowSize in sequenceB\n",
    "            substringB = sequenceB[indexB:indexB+windowSize]\n",
    "            if substringA == substringB:\n",
    "                xList.append(indexA)\n",
    "                yList.append(indexB)\n",
    "\n",
    "    # Use matplotlib to create a scatter plot of all of the points where the sequences match.\n",
    "    pyplot.scatter(xList, yList)\n",
    "    pyplot.axis([0, maxIndexA, 0, maxIndexB])\n",
    "    pyplot.show()\n",
    "\n",
    "def globalAlignment(s, t, substitutionMatrix, indelPenalty):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "\n",
    "    # TODO s is the column, t is the row\n",
    "\n",
    "    # Set up the scores and paths lists\n",
    "    scores = []\n",
    "    paths = []\n",
    "    for _ in range(len(s)+1):\n",
    "        thisScoresRow = []\n",
    "        thisPathsRow = []\n",
    "        for _ in range(len(t)+1):\n",
    "            thisScoresRow.append(0)\n",
    "            thisPathsRow.append(\"\")\n",
    "        scores.append(thisScoresRow)\n",
    "        paths.append(thisPathsRow)\n",
    "\n",
    "    # The top left corner should be initialized with a score of 0 and a path of \"\"\n",
    "\n",
    "    # Compute the top row of scores, excluding the top left corner\n",
    "    for tPos in range(1, len(t)+1):\n",
    "        scores[0][tPos] = scores[0][tPos-1] + indelPenalty\n",
    "        paths[0][tPos] = \"L\"\n",
    "\n",
    "    # Compute the left row of scores, excluding the top left corner\n",
    "    for sPos in range(1, len(s)+1):\n",
    "        scores[sPos][0] = scores[sPos-1][0] + indelPenalty\n",
    "        paths[sPos][0] = \"U\"\n",
    "\n",
    "    # Compute the rest of the scores, based on the previous values in the scores 2D list\n",
    "    # We'll calculate the first row, from left to right, followed by the 2nd row, and so on.\n",
    "    for sPos in range(1, len(s)+1):\n",
    "        for tPos in range(1, len(t)+1):\n",
    "            scoreFromLeft = scores[sPos][tPos-1] + indelPenalty\n",
    "            scoreFromUp = scores[sPos-1][tPos] + indelPenalty\n",
    "            sChar = s[sPos-1]\n",
    "            tChar = t[tPos-1]\n",
    "            scoreFromDiagonal = scores[sPos-1][tPos-1] + substitutionMatrix[sChar,tChar]\n",
    "\n",
    "            # In the case of a tie, we prefer mismatches/matches to indels.\n",
    "            if scoreFromDiagonal >= scoreFromLeft and scoreFromDiagonal >= scoreFromUp:\n",
    "                # scoreFromDiagonal is biggest\n",
    "                scores[sPos][tPos] = scoreFromDiagonal\n",
    "                paths[sPos][tPos] = \"D\"\n",
    "            elif scoreFromLeft >= scoreFromUp:\n",
    "                # scoreFromLeft is biggest\n",
    "                scores[sPos][tPos] = scoreFromLeft\n",
    "                paths[sPos][tPos] = \"L\"\n",
    "            else:\n",
    "                # scoreFromUp is biggest\n",
    "                scores[sPos][tPos] = scoreFromUp\n",
    "                paths[sPos][tPos] = \"U\"\n",
    "\n",
    "    # Work backward through the paths 2D list to reconstruct the alignment.\n",
    "    sAlignment = \"\"\n",
    "    tAlignment = \"\"\n",
    "    sPos = len(s)\n",
    "    tPos = len(t)\n",
    "    while not (sPos == 0 and tPos == 0):\n",
    "        if paths[sPos][tPos] == \"D\":\n",
    "            sAlignment = s[sPos-1] + sAlignment\n",
    "            tAlignment = t[tPos-1] + tAlignment\n",
    "            sPos -= 1\n",
    "            tPos -= 1\n",
    "        elif paths[sPos][tPos] == \"L\":\n",
    "            sAlignment = \"-\" + sAlignment\n",
    "            tAlignment = t[tPos-1] + tAlignment\n",
    "            tPos -= 1\n",
    "        else: # paths[sPos][tPos] == \"U\"\n",
    "            sAlignment = s[sPos-1] + sAlignment\n",
    "            tAlignment = \"-\" + tAlignment\n",
    "            sPos -= 1\n",
    "\n",
    "    maxScore = scores[len(s)][len(t)]\n",
    "\n",
    "    return maxScore, [sAlignment, tAlignment]\n",
    "\n",
    "def localAlignment(s, t, substitutionMatrix, indelPenalty):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "\n",
    "    # TODO s is the column, t is the row\n",
    "\n",
    "    # Set up the scores and paths lists\n",
    "    scores = []\n",
    "    paths = []\n",
    "    for _ in range(len(s)+1):\n",
    "        thisScoresRow = []\n",
    "        thisPathsRow = []\n",
    "        for _ in range(len(t)+1):\n",
    "            thisScoresRow.append(0)\n",
    "            thisPathsRow.append(\"\")\n",
    "        scores.append(thisScoresRow)\n",
    "        paths.append(thisPathsRow)\n",
    "\n",
    "    # The top left corner should be initialized with a score of 0 and a path of \"\"\n",
    "\n",
    "    # A blank entry in the paths 2D list indicates the start point for local alignment\n",
    "\n",
    "    # Compute the rest of the scores, based on the previous values in the scores 2D list\n",
    "    # We'll calculate the first row, from left to right, followed by the 2nd row, and so on.\n",
    "    for sPos in range(1, len(s)+1):\n",
    "        for tPos in range(1, len(t)+1):\n",
    "            scoreFromStartingNewSequence = 0\n",
    "            scoreFromLeft = scores[sPos][tPos-1] + indelPenalty\n",
    "            scoreFromUp = scores[sPos-1][tPos] + indelPenalty\n",
    "            sChar = s[sPos-1]\n",
    "            tChar = t[tPos-1]\n",
    "            scoreFromDiagonal = scores[sPos-1][tPos-1] + substitutionMatrix[sChar,tChar]\n",
    "\n",
    "            # In the case of a tie, we prefer mismatches/matches to indels.\n",
    "            if scoreFromDiagonal >= scoreFromLeft and scoreFromDiagonal >= scoreFromUp and scoreFromDiagonal >= scoreFromStartingNewSequence:\n",
    "                # scoreFromDiagonal is biggest\n",
    "                scores[sPos][tPos] = scoreFromDiagonal\n",
    "                paths[sPos][tPos] = \"D\"\n",
    "            elif scoreFromLeft >= scoreFromUp and scoreFromLeft >= scoreFromStartingNewSequence:\n",
    "                # scoreFromLeft is biggest\n",
    "                scores[sPos][tPos] = scoreFromLeft\n",
    "                paths[sPos][tPos] = \"L\"\n",
    "            elif scoreFromUp >= scoreFromStartingNewSequence:\n",
    "                # scoreFromUp is biggest\n",
    "                scores[sPos][tPos] = scoreFromUp\n",
    "                paths[sPos][tPos] = \"U\"\n",
    "            else:\n",
    "                # scoreFromStartingNewSequence is biggest, so let's start a new sequence\n",
    "                scores[sPos][tPos] = scoreFromStartingNewSequence\n",
    "                paths[sPos][tPos] = \"\"\n",
    "\n",
    "    # Find the point in the scores 2D list with the highest score:\n",
    "    highestScore = -1\n",
    "    highestScoreSPos = -1\n",
    "    highestScoreTPos = -1\n",
    "    for sPos in range(0, len(s)+1):\n",
    "        for tPos in range(0, len(t)+1):\n",
    "            if scores[sPos][tPos] > highestScore:\n",
    "                highestScore = scores[sPos][tPos]\n",
    "                highestScoreSPos = sPos\n",
    "                highestScoreTPos = tPos\n",
    "\n",
    "    # Work backward through the paths 2D list to reconstruct the alignment.\n",
    "    sAlignment = \"\"\n",
    "    tAlignment = \"\"\n",
    "    sPos = highestScoreSPos\n",
    "    tPos = highestScoreTPos\n",
    "    while not (paths[sPos][tPos] == \"\"):\n",
    "        if paths[sPos][tPos] == \"D\":\n",
    "            sAlignment = s[sPos-1] + sAlignment\n",
    "            tAlignment = t[tPos-1] + tAlignment\n",
    "            sPos -= 1\n",
    "            tPos -= 1\n",
    "        elif paths[sPos][tPos] == \"L\":\n",
    "            sAlignment = \"-\" + sAlignment\n",
    "            tAlignment = t[tPos-1] + tAlignment\n",
    "            tPos -= 1\n",
    "        else: # paths[sPos][tPos] == \"U\"\n",
    "            sAlignment = s[sPos-1] + sAlignment\n",
    "            tAlignment = \"-\" + tAlignment\n",
    "            sPos -= 1\n",
    "\n",
    "    startPos = (sPos, tPos)\n",
    "    endPos = (highestScoreSPos, highestScoreTPos)\n",
    "    return highestScore, [sAlignment, tAlignment], startPos, endPos\n",
    "\n",
    "def shuffled(string):\n",
    "    \"\"\"\n",
    "    Source: https://stackoverflow.com/questions/2668312/shuffle-string-in-python/2668366\n",
    "    \"\"\"\n",
    "    listOfCharacters = list(string)\n",
    "    random.shuffle(listOfCharacters)\n",
    "    result = ''.join(listOfCharacters)\n",
    "\n",
    "    return result\n",
    "\n",
    "def runAlignmentAndFindPValue(s, t, substitutionMatrix, indelPenalty, alignmentFunc):\n",
    "    result = alignmentFunc(s, t, substitutionMatrix, indelPenalty)\n",
    "    realScore = result[0]\n",
    "    print(f\"realScore {realScore}\", file=sys.stderr)\n",
    "\n",
    "    countOfRandomScoresBetterThanRealScore = 0\n",
    "    for i in range(200):\n",
    "        s = shuffled(s)\n",
    "        result = alignmentFunc(s, t, substitutionMatrix, indelPenalty)\n",
    "        score = result[0]\n",
    "        print(f\"score {score}\", file=sys.stderr)\n",
    "        if score >= realScore:\n",
    "            countOfRandomScoresBetterThanRealScore += 1\n",
    "\n",
    "    print(f\"p-value of alignment: {countOfRandomScoresBetterThanRealScore}/{200} = {countOfRandomScoresBetterThanRealScore/200.0}\", file=sys.stderr)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "realScore 1100.0\n",
      "score -183.0\n",
      "score -198.0\n",
      "score -183.0\n",
      "score -162.0\n",
      "score -186.0\n",
      "score -108.0\n",
      "score -183.0\n",
      "score -224.0\n",
      "score -191.0\n",
      "score -229.0\n",
      "score -166.0\n",
      "score -161.0\n",
      "score -176.0\n",
      "score -238.0\n",
      "score -140.0\n",
      "score -178.0\n",
      "score -162.0\n",
      "score -254.0\n",
      "score -244.0\n",
      "score -198.0\n",
      "score -235.0\n",
      "score -199.0\n",
      "score -158.0\n",
      "score -190.0\n",
      "score -221.0\n",
      "score -213.0\n",
      "score -233.0\n",
      "score -151.0\n",
      "score -167.0\n",
      "score -194.0\n",
      "score -264.0\n",
      "score -187.0\n",
      "score -185.0\n",
      "score -208.0\n",
      "score -191.0\n",
      "score -148.0\n",
      "score -204.0\n",
      "score -208.0\n",
      "score -242.0\n",
      "score -146.0\n",
      "score -123.0\n",
      "score -155.0\n",
      "score -141.0\n",
      "score -179.0\n",
      "score -124.0\n",
      "score -226.0\n",
      "score -204.0\n",
      "score -171.0\n",
      "score -118.0\n",
      "score -202.0\n",
      "score -221.0\n",
      "score -184.0\n",
      "score -147.0\n",
      "score -163.0\n",
      "score -173.0\n",
      "score -212.0\n",
      "score -202.0\n",
      "score -156.0\n",
      "score -207.0\n",
      "score -164.0\n",
      "score -169.0\n",
      "score -188.0\n",
      "score -222.0\n",
      "score -208.0\n",
      "score -206.0\n",
      "score -180.0\n",
      "score -188.0\n",
      "score -273.0\n",
      "score -141.0\n",
      "score -212.0\n",
      "score -157.0\n",
      "score -237.0\n",
      "score -166.0\n",
      "score -186.0\n",
      "score -172.0\n",
      "score -232.0\n",
      "score -164.0\n",
      "score -210.0\n",
      "score -135.0\n",
      "score -194.0\n",
      "score -156.0\n",
      "score -165.0\n",
      "score -145.0\n",
      "score -157.0\n",
      "score -230.0\n",
      "score -245.0\n",
      "score -189.0\n",
      "score -214.0\n",
      "score -135.0\n",
      "score -174.0\n",
      "score -273.0\n",
      "score -151.0\n",
      "score -241.0\n",
      "score -243.0\n",
      "score -204.0\n",
      "score -146.0\n",
      "score -176.0\n",
      "score -260.0\n",
      "score -171.0\n",
      "score -166.0\n",
      "score -217.0\n",
      "score -148.0\n",
      "score -188.0\n",
      "score -189.0\n",
      "score -142.0\n",
      "score -157.0\n",
      "score -174.0\n",
      "score -193.0\n",
      "score -201.0\n",
      "score -208.0\n",
      "score -192.0\n",
      "score -156.0\n",
      "score -192.0\n",
      "score -194.0\n",
      "score -192.0\n",
      "score -173.0\n",
      "score -245.0\n",
      "score -232.0\n",
      "score -213.0\n",
      "score -145.0\n",
      "score -177.0\n",
      "score -268.0\n",
      "score -138.0\n",
      "score -229.0\n",
      "score -164.0\n",
      "score -222.0\n",
      "score -206.0\n",
      "score -214.0\n",
      "score -278.0\n",
      "score -152.0\n",
      "score -225.0\n",
      "score -258.0\n",
      "score -209.0\n",
      "score -181.0\n",
      "score -180.0\n",
      "score -170.0\n",
      "score -293.0\n",
      "score -181.0\n",
      "score -267.0\n",
      "score -236.0\n",
      "score -185.0\n",
      "score -239.0\n",
      "score -230.0\n",
      "score -196.0\n",
      "score -161.0\n",
      "score -199.0\n",
      "score -220.0\n",
      "score -196.0\n",
      "score -186.0\n",
      "score -181.0\n",
      "score -177.0\n",
      "score -249.0\n",
      "score -182.0\n",
      "score -209.0\n",
      "score -214.0\n",
      "score -232.0\n",
      "score -246.0\n",
      "score -181.0\n",
      "score -183.0\n",
      "score -245.0\n",
      "score -254.0\n",
      "score -254.0\n",
      "score -202.0\n",
      "score -191.0\n",
      "score -165.0\n",
      "score -209.0\n",
      "score -219.0\n",
      "score -157.0\n",
      "score -196.0\n",
      "score -272.0\n",
      "score -127.0\n",
      "score -204.0\n",
      "score -174.0\n",
      "score -220.0\n",
      "score -197.0\n",
      "score -219.0\n",
      "score -203.0\n",
      "score -146.0\n",
      "score -219.0\n",
      "score -214.0\n",
      "score -188.0\n",
      "score -189.0\n",
      "score -186.0\n",
      "score -176.0\n",
      "score -101.0\n",
      "score -244.0\n",
      "score -223.0\n",
      "score -152.0\n",
      "score -249.0\n",
      "score -182.0\n",
      "score -218.0\n",
      "score -185.0\n",
      "score -141.0\n",
      "score -198.0\n",
      "score -200.0\n",
      "score -197.0\n",
      "score -184.0\n",
      "score -185.0\n",
      "score -220.0\n",
      "score -236.0\n",
      "p-value of alignment: 0/200 = 0.0\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'startPos' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-f03dc1d3cfdd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpax6Alignment\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 41\u001b[0;31m \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     42\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-2-f03dc1d3cfdd>\u001b[0m in \u001b[0;36mmain\u001b[0;34m()\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     36\u001b[0m     \u001b[0;31m#score, [eyelessAlignment, pax6Alignment] = compbio.globalAlignment(eyelessmRNA, pax6mRNA, basicDNASubstitutionMatrix, indelPenalty)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 37\u001b[0;31m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"mRNA local alignment score: {score} startPos: {startPos} endPos: {endPos}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     38\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0meyelessAlignment\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpax6Alignment\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'startPos' is not defined"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "from Bio.Align import substitution_matrices\n",
    "\n",
    "import compbio\n",
    "\n",
    "def main():\n",
    "    Entrez.email = \"reiche_n1@denison.edu\"\n",
    "\n",
    "    eyelessProtein = compbio.getSequencesFromNCBI(\"protein\", \"NP_524628.2\")[0]\n",
    "    eyelessmRNA = compbio.getSequencesFromNCBI(\"nucleotide\", \"NM_079889.3\")[0]\n",
    "    eyelessGenomeDNA = compbio.getSequencesFromFile(\"eyeless_genomic_sequence.fasta\")[0]\n",
    "\n",
    "    pax6Protein = compbio.getSequencesFromNCBI(\"protein\", \"1902328A\")[0]\n",
    "    pax6mRNA = compbio.getSequencesFromNCBI(\"nucleotide\", \"NM_000280.4\")[0]\n",
    "    pax6GenomeDNA = compbio.getSequencesFromFile(\"PAX6_genomic_sequence.fasta\")[0]\n",
    "\n",
    "    #compbio.dotplot(eyelessProtein, pax6Protein, 4)\n",
    "    #compbio.dotplot(eyelessmRNA, pax6mRNA, 8)\n",
    "\n",
    "    indelPenalty = -1\n",
    "    basicDNASubstitutionMatrix = {\n",
    "            ('A','A'):  1, ('A','C'): -1, ('A','G'): -1, ('A','T'): -1,\n",
    "            ('C','A'): -1, ('C','C'):  1, ('C','G'): -1, ('C','T'): -1,\n",
    "            ('G','A'): -1, ('G','C'): -1, ('G','G'):  1, ('G','T'): -1,\n",
    "            ('T','A'): -1, ('T','C'): -1, ('T','G'): -1, ('T','T'):  1,\n",
    "    }\n",
    "    #nuc = substitution_matrices.load(\"NUC.4.4\")\n",
    "    #score, [eyelessAlignment, pax6Alignment], startPos, endPos = compbio.localAlignment(eyelessmRNA, pax6mRNA, nuc, -8)\n",
    "    #blosum = substitution_matrices.load(\"BLOSUM80\")\n",
    "    #score, [eyelessAlignment, pax6Alignment], startPos, endPos = compbio.localAlignment(eyelessProtein, pax6Protein, blosum, -4)\n",
    "    #compbio.runAlignmentAndFindPValue(eyelessProtein, pax6Protein, blosum, -4, compbio.localAlignment)\n",
    "    blosum = substitution_matrices.load(\"BLOSUM80\")\n",
    "    score, [eyelessAlignment, pax6Alignment] = compbio.globalAlignment(eyelessProtein, pax6Protein, blosum, -4)\n",
    "    compbio.runAlignmentAndFindPValue(eyelessProtein, pax6Protein, blosum, -4, compbio.globalAlignment)\n",
    "\n",
    "    #score, [eyelessAlignment, pax6Alignment] = compbio.globalAlignment(eyelessmRNA, pax6mRNA, basicDNASubstitutionMatrix, indelPenalty)\n",
    "    print(f\"mRNA local alignment score: {score} startPos: {startPos} endPos: {endPos}\")\n",
    "    print(eyelessAlignment)\n",
    "    print(pax6Alignment)\n",
    "\n",
    "main()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
