{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Highest N50 value: 25419 when K= 35 the set of contigs has 16 contigs each has a length of [43, 46, 232, 607, 1147, 1175, 1686, 2898, 5621, 10674, 11784, 11863, 14378, 25419, 26425, 48930]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\nprint(num_of_contigs,contig_max_len)\\nprint(min(num_of_contigs),num_of_contigs.index(min(num_of_contigs)))\\nprint(max(contig_max_len), contig_max_len.index(max(contig_max_len)))\\n'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import math\n",
    "import random\n",
    "from Bio import Entrez, SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import queue\n",
    "\n",
    "def readpairs(name,k):\n",
    "    \"\"\"\n",
    "    Read the FASTA file and return a list of reads and a list of k-mers from the reads\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    name : FASTA file of bacterial genome\n",
    "    k : (integer) length of reads\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    A list of reads and a list of k-mers from the read\n",
    "\n",
    "    \"\"\"\n",
    "    kmers = []                                                  #empty list to store k-mers\n",
    "    sequences = []                                              #empty list to stores reads\n",
    "    for record in SeqIO.parse( (name + '.fasta'),'fasta'):      #iterate over FASTA file and parse function separates read labels and read-pairs\n",
    "        sequences.append(str(record.seq))                       #add reads to reads list\n",
    "    for i in range( len(sequences) ):                           #iterate over index of reads list to generate k-mers\n",
    "        for j in range( len(sequences[i])-k+1 ) :\n",
    "            kmer = sequences[i][j:j+k]                          \n",
    "            kmers.append(kmer)                                  #add k-mers to k-mers liist\n",
    "    return (sequences,kmers)                                    #return lists for reads and k-mers\n",
    "\n",
    "\n",
    "def create_deBruijn(kmers):\n",
    "    \"\"\"\n",
    "    Create a de Bruijn graph from a set of k-mers\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    kmers : (string) pattern of bases from the readsof the bacteria genome\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Return a graph dictionary, a dictionary for in-degrees of each node, and a \n",
    "    dictionary for out-degress of each node\n",
    "\n",
    "    \"\"\"\n",
    "    count = {}                                    #empty dictionary to record k-mers\n",
    "    graph = {}                                    #empty dictionary to record nodes and list of nodes connected by a directed edge\n",
    "    in_degrees = {}                               #empty dictionary to record in-degrees of nodes\n",
    "    out_degrees = {}                              #empty dictionary to record out-degrees of nodes\n",
    "    for i in range(len(kmers)):                   #for loop iterates over list of k-mers  \n",
    "        if kmers[i] not in count:                 #if k-mers index not in k-mers dictionary, then node prefix is in the k-mer backward one position  \n",
    "            node_p = kmers[i][:-1]\n",
    "            if node_p not in graph:               #if node prefix not in graph dictionary then node prefix is in empty list      \n",
    "                graph[node_p] = []\n",
    "                in_degrees[node_p] = 0            #in-degree of node prefix equals 0 for the node  \n",
    "                out_degrees[node_p] = 0           #out-degree of node prefix equals 0 for the node  \n",
    "            node_s = kmers[i][1:]                 #node suffix equals the k-mers index position\n",
    "            if node_s not in graph:               #if node suffix not in graph dictionary then node suffix is in empty list  \n",
    "                graph[node_s] = []\n",
    "                in_degrees[node_s] = 0            #in-degree of node suffix equals 0\n",
    "                out_degrees[node_s] = 0           #out-degree of node suffic equals 0  \n",
    "            graph[node_p].append(node_s)          #add node prefix to node suffix list  \n",
    "            in_degrees[node_s] += 1               #in_degree of node suffix add one value  \n",
    "            out_degrees[node_p] += 1              #out-degree of node prefix add one value\n",
    "            count[kmers[i]] = 0                   #k-mers index in dictionary equals 0  \n",
    "    return (graph,in_degrees,out_degrees)         #return graph, in-degree, and our-degree dictionaries   \n",
    "\n",
    "\n",
    "def GetMaximalNonbranchingPaths(graph, inDegrees, outDegrees):\n",
    "    \"\"\"\n",
    "    Find all maximal non-branching paths in the de Bruijn graph. Which are paths with internal\n",
    "    nodes that have in-degree and out-degree equal to one except for the starting and ending node\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    graph : dictionary with keys as the nodes ((k-1)mers) and values as a list of directed edge\n",
    "            that connects nodes and keys\n",
    "    inDegrees : dictionary of keys as nodes and values as the corresponding number of in-degrees\n",
    "    outDegrees : dictionary of keys as nodes and values as the corresponding number of out-degrees\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    paths: list of maximal non-branching paths in the de Bruijn graph, consist of sets of k-mers\n",
    "\n",
    "    \"\"\"\n",
    "    paths = []                                              #empty list to store maximal non-branching paths\n",
    "    nodes11 = {}                                            #empty dictionary to store set of nodes with in-degree and out-degree both equal to 1\n",
    "    for id in graph:                                        #for loop iterate over index of graph\n",
    "        if inDegrees[id] == 1 and outDegrees[id] == 1:      #if in-degree and out-degree of index in graph equals 1\n",
    "            nodes11[id] = 0                                 #then the values of nodes11 dictionary equals 0\n",
    "    \n",
    "    for start in graph:                                     #iterate over each node start in graph    \n",
    "        if start not in nodes11:                            #if node start is not in nodes11 dictionary\n",
    "            for current in graph[start]:                    #iterate over each current node in adjacent list of start\n",
    "                path = [start,current]                      #path equals node start and node current\n",
    "                while (current in nodes11):                 #while current node in is nodes11 dictionary\n",
    "                    next_ = graph[current][0]               #set variable for node next to current node\n",
    "                    path.append(next_)                      #append next node to paths list\n",
    "                    graph[current].remove(next_)            #remove next node from current node list\n",
    "                    current = next_                         #set current node as next node\n",
    "                paths.append(path)                          #append paths to list of paths\n",
    "    \n",
    "    for start in nodes11:                                   #iterate over each start node in the nodes11 dictionary\n",
    "        if (graph[start] != []):                            #if start node does not have an adjacent node then follow conditions\n",
    "            path = [start]                                  #set path equal to start node\n",
    "            current = graph[start][0]                       #set current node to the node adjacent to start node\n",
    "            while (current != start):                       #while current node doesn't equal start node add current node to path\n",
    "                path.append(current)                    \n",
    "                next_ = graph[current][0]                   #set next variable equal to the node adjacent to the current node\n",
    "                graph[current].remove(next_)                #remove current and next node from the graph\n",
    "                current = next_                             #set current node equal to the next node\n",
    "            paths.append(path)                              #add path to paths list\n",
    "    \n",
    "    return paths                                            #return paths list\n",
    "\n",
    "def constructContigs(paths):\n",
    "    \"\"\"\n",
    "    Function creates contigs (overlapping DNA bases) from a set of paths\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    paths : list of kmers that is the maximal non-branching path in the de Bruijn graph\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Returns a list of contigs\n",
    "\n",
    "    \"\"\"\n",
    "    contigs = []                                #empty list to store contigs\n",
    "    for i in range(len(paths)):                 #iterate over index of paths list\n",
    "        path = paths[i]                         #set variable path equal to index of paths list\n",
    "        contig = path[0]                        #set variable contig equal to first index of path\n",
    "        for j in range(1,len(path)):            #iterate over next index of path\n",
    "            contig = contig + path[j][-1]       #set contig equal to sum of contig and path index minus 1\n",
    "        contigs.append(contig)                  #add contig to contigs list\n",
    "    return (contigs)                            #return contigs as a list \n",
    "\n",
    "def calculateN50(l):\n",
    "    \"\"\"\n",
    "    Function creates contigs from the set of reads with different k values. Use N50 to measure\n",
    "    the qualuty of the assembly of contigs\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    l = (list) set of reads\n",
    "    \n",
    "    Returns\n",
    "    ---------\n",
    "    Returns N50 value: maximal contig length that the contigs greater than or equal to the length\n",
    "    that is equal to half of the sun of lengths of contigs\n",
    "    \"\"\"\n",
    "    \n",
    "    middle = sum(l)//2     #set middle variable to calculate half of sum of the lengths of all contigs\n",
    "    l.sort()               #sort the list of reads in order from minimum to maximum  \n",
    "    i = -1                 #set index of congs equal to -1         \n",
    "    N50 = l[i]             #set N50 variable equal to reads list\n",
    "    count = N50            #set count varaible equal to N50\n",
    "    while count < middle:  #while count is less than 50% of the total lengths of contigs\n",
    "        i -= 1             #index equals minus 1 value\n",
    "        N50 = l[i]         #set N50 equal to index of reads list\n",
    "        count += N50       #set count equal to add value of N50\n",
    "    return N50             #return N50 value\n",
    "\n",
    "def orderingContigs(contigs, reads):\n",
    "    order_graph = {}\n",
    "    indegree = {}\n",
    "    final_contigs = []\n",
    "    for i in range( len(contigs) ):  # iterate over the length of contigs\n",
    "        current = contigs[i]          # set the ith contig as current\n",
    "        for j in range( 0, len(reads), 2 ):   #the second index of the read label\n",
    "            if reads[j] in current:       \n",
    "                if current not in order_graph:   # if ith contig is not in order_graph dictionary\n",
    "                    order_graph[current] = []     # set the ith contig as key and assign it with an empty list\n",
    "                if current not in indegree:       #if ith contig is not in indegree dictionary\n",
    "                    indegree[current] = 0         # set the indgree of the contig as 0\n",
    "                for x in range( len(contigs) ):\n",
    "                    if i != x:       \n",
    "                        if reads[j+1] in contigs[x]:   # if the reads is not in contigs \n",
    "                            follow = contigs[x]        # set the contig as follow\n",
    "                            if follow not in indegree:\n",
    "                                indegree[follow] = 0     #set the indegree of follow as 0\n",
    "                            order_graph[current].append(contigs[x])     #combine contigs\n",
    "                            indegree[follow] += 1        #add indegree 1 to the follow variable\n",
    "    \n",
    "    #order = queue.Queue()\n",
    "    order = []\n",
    "    for key in order_graph:  # iterate over the order_graph dictionary we set before\n",
    "        if indegree[key] == 0:  # if the indegree of the contig is 0\n",
    "            #order.put(id)\n",
    "            order.append(key)  # append the first contig that with the indegree of 1 to order list\n",
    "\n",
    "    while (len(order)!=0): #not(order.empty())\n",
    "        key = order.pop(-1)     \n",
    "        final_contigs.append(key)       \n",
    "        if key in order_graph:\n",
    "            succeessor = order_graph[key] \n",
    "        if (len(succeessor) != 0 ):         # if the indegree of the key(contig) is not 0\n",
    "            for i in range( len(succeessor) ):\n",
    "                node = succeessor[i]      # node is the contig \n",
    "                indegree[node] -= 1     # the indegree of node minus 1\n",
    "                if (indegree[node] == 0):\n",
    "                    order.append(node)    # append all nodes that with indegree of 0 into the order\n",
    "    #print(len(final_contigs))\n",
    "    fasta_seq = []\n",
    "    for i in range(len(final_contigs)):\n",
    "        contig = final_contigs[i]       # set the contig as each key in final list\n",
    "        record = SeqRecord(Seq(contig), \"contig%i\" % (i + 1), \"\", \"\")    # build the fasta file for contigs\n",
    "        fasta_seq.append(record) # append the record to fasta file\n",
    "    j = 1\n",
    "    for i in range(len(contigs)):\n",
    "        if contigs[i] not in final_contigs:\n",
    "            contig = contigs[i]    # set the contig as each key in contigs liset\n",
    "            record = SeqRecord(Seq(contig), \"contig%i\" % ( len(final_contigs) +j ), \"\", \"\") # build the fasta file for contigs\n",
    "            fasta_seq.append(record)     # append the record to fasta file\n",
    "            j+=1\n",
    "    SeqIO.write(fasta_seq, \"lab3_result.fasta\", \"fasta\") #write out fasta file\n",
    "\n",
    "\n",
    "def main():\n",
    "    \"\"\"\n",
    "    Functions are called when given the FASTA file for C_ruddii bacterium.\n",
    "    Returns a list of reads and a list of k-mers, graph dictionary, in-degree dictionary, out-degree dictionary,\n",
    "    list of contigs, best N50, best K value, number of contigsm, and the length of each contig.\n",
    "\n",
    "    \"\"\"\n",
    "    num_of_contigs = []         #empty list to store number of contigs\n",
    "    contig_max_len = []         #empty list to store maximum length of contig\n",
    "    Best_N50 = 0                \n",
    "    Best_K = 20\n",
    "    Best_set_contig_len = []    #empty list to store best set of contig length\n",
    "    num_of_contigs = 0\n",
    "    Best_contigs = []\n",
    "    for j in range(20,41):                        #between 20 and 41 as values for k, variables are assigned to call functions\n",
    "        reads,kmers = readpairs('C_ruddii',j)\n",
    "        graph,in_degrees,out_degrees = create_deBruijn(kmers)\n",
    "        paths = GetMaximalNonbranchingPaths(graph, in_degrees, out_degrees)\n",
    "        contigs = constructContigs(paths)\n",
    "        a = []                                   #empty list to store length of contigs       \n",
    "        for i in range( len(contigs) ):          #iterate over length of contigs and add them to list a\n",
    "            a.append(len(contigs[i]) )\n",
    "        N50 = calculateN50(a)                    #calculate N50 for maximal contig length \n",
    "        if N50 > Best_N50:\n",
    "            Best_N50 = N50\n",
    "            Best_K = 20+i\n",
    "            num_of_contigs = len(contigs)\n",
    "            Best_set_contig_len = a\n",
    "            Best_contigs = contigs\n",
    "            \n",
    "    print('Highest N50 value:',Best_N50,'when K=', Best_K,'the set of contigs has',num_of_contigs, 'contigs each has a length of', Best_set_contig_len)\n",
    "    #print(paths)\n",
    "    orderingContigs(Best_contigs,reads)\n",
    "    \n",
    "main()\n",
    "\n",
    "\"\"\"\n",
    "print(num_of_contigs,contig_max_len)\n",
    "print(min(num_of_contigs),num_of_contigs.index(min(num_of_contigs)))\n",
    "print(max(contig_max_len), contig_max_len.index(max(contig_max_len)))\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
